File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/prefer-equality-matcher.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

const expectSuggestions = (output: (equalityMatcher: string) => string) =>
  ['toBe', 'toEqual', 'toStrictEqual'].map(equalityMatcher => ({
    messageId: 'suggestEqualityMatcher',
    data: { equalityMatcher },
    output: output(equalityMatcher),
  }));

ruleTester.run('prefer-equality-matcher', {} as never, {
  valid: [
    // ===
    { code: 'expect.hasAssertions' },
    { code: 'expect.hasAssertions()' },
    { code: 'expect.assertions(1)' },
    { code: 'expect(a == 1).toBe(true)' },
    { code: 'expect(1 == a).toBe(true)' },
    { code: 'expect(a == b).toBe(true)' },
    // !==
    { code: 'expect(a != 1).toBe(true)' },
    { code: 'expect(1 != a).toBe(true)' },
    { code: 'expect(a != b).toBe(true)' },
  ],
  invalid: [
    // ===
    {
      code: 'expect(a === b).toBe(true);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).${equalityMatcher}(b);`,
          ),
          column: 17,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a === b,).toBe(true,);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a,).${equalityMatcher}(b,);`,
          ),
          column: 18,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a === b).toBe(false);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).not.${equalityMatcher}(b);`,
          ),
          column: 17,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a === b).resolves.toBe(true);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).resolves.${equalityMatcher}(b);`,
          ),
          column: 26,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a === b).resolves.toBe(false);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).resolves.not.${equalityMatcher}(b);`,
          ),
          column: 26,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a === b).not.toBe(true);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).not.${equalityMatcher}(b);`,
          ),
          column: 21,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a === b).not.toBe(false);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).${equalityMatcher}(b);`,
          ),
          column: 21,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a === b).resolves.not.toBe(true);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).resolves.not.${equalityMatcher}(b);`,
          ),
          column: 30,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a === b).resolves.not.toBe(false);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).resolves.${equalityMatcher}(b);`,
          ),
          column: 30,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a === b)["resolves"].not.toBe(false);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).resolves.${equalityMatcher}(b);`,
          ),
          column: 33,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a === b)["resolves"]["not"]["toBe"](false);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).resolves.${equalityMatcher}(b);`,
          ),
          column: 36,
          line: 1,
        },
      ],
    },
    // !==
    {
      code: 'expect(a !== b).toBe(true);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).not.${equalityMatcher}(b);`,
          ),
          column: 17,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a !== b).toBe(false);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).${equalityMatcher}(b);`,
          ),
          column: 17,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a !== b).resolves.toBe(true);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).resolves.not.${equalityMatcher}(b);`,
          ),
          column: 26,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a !== b).resolves.toBe(false);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).resolves.${equalityMatcher}(b);`,
          ),
          column: 26,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a !== b).not.toBe(true);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).${equalityMatcher}(b);`,
          ),
          column: 21,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a !== b).not.toBe(false);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).not.${equalityMatcher}(b);`,
          ),
          column: 21,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a !== b).resolves.not.toBe(true);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).resolves.${equalityMatcher}(b);`,
          ),
          column: 30,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a !== b).resolves.not.toBe(false);',
      errors: [
        {
          messageId: 'useEqualityMatcher',
          suggestions: expectSuggestions(
            equalityMatcher => `expect(a).resolves.not.${equalityMatcher}(b);`,
          ),
          column: 30,
          line: 1,
        },
      ],
    },
  ],
});

File: internal/plugins/jest/rules/prefer_equality_matcher/prefer_equality_matcher_test.go
package prefer_equality_matcher_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/prefer_equality_matcher"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func expectSuggestions(output func(equalityMatcher string) string) []rule_tester.InvalidTestCaseSuggestion {
	matchers := []string{"toBe", "toEqual", "toStrictEqual"}
	suggestions := make([]rule_tester.InvalidTestCaseSuggestion, len(matchers))
	for i, matcher := range matchers {
		suggestions[i] = rule_tester.InvalidTestCaseSuggestion{
			MessageId: "suggestEqualityMatcher",
			Output:    output(matcher),
		}
	}
	return suggestions
}

func TestPreferEqualityMatcherRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&prefer_equality_matcher.PreferEqualityMatcherRule,
		[]rule_tester.ValidTestCase{
			// ===
			{Code: `expect.hasAssertions`},
			{Code: `expect.hasAssertions()`},
			{Code: `expect.assertions(1)`},
			{Code: `expect(a == 1).toBe(true)`},
			{Code: `expect(1 == a).toBe(true)`},
			{Code: `expect(a == b).toBe(true)`},
			{Code: `expect((a === b)).toBe(true)`},
			{Code: `expect((a === b) as boolean).toBe(true)`},
			{Code: `expect(a === b).toBe((true))`},
			{Code: `expect(a === b).toBe(true as boolean)`},
			// !==
			{Code: `expect(a != 1).toBe(true)`},
			{Code: `expect(1 != a).toBe(true)`},
			{Code: `expect(a != b).toBe(true)`},
			{Code: `expect((a !== b)).toBe(true)`},
		},
		[]rule_tester.InvalidTestCase{
			// ===
			{
				Code: `expect(a === b).toBe(true);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    17,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a === b,).toBe(true,);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    18,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a,).` + equalityMatcher + `(b,);`
						}),
					},
				},
			},
			{
				Code: `expect(a === b).toBe(false);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    17,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).not.` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a === b).resolves.toBe(true);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    26,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).resolves.` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a === b).resolves.toBe(false);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    26,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).resolves.not.` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a === b).not.toBe(true);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    21,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).not.` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a === b).not.toBe(false);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    21,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a === b).resolves.not.toBe(true);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    30,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).resolves.not.` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a === b).resolves.not.toBe(false);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    30,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).resolves.` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a === b)["resolves"].not.toBe(false);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    33,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).resolves.` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a === b)["resolves"]["not"]["toBe"](false);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    36,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).resolves.` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			// !==
			{
				Code: `expect(a !== b).toBe(true);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    17,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).not.` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a !== b).toBe(false);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    17,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a !== b).resolves.toBe(true);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    26,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).resolves.not.` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a !== b).resolves.toBe(false);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    26,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).resolves.` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a !== b).not.toBe(true);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    21,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a !== b).not.toBe(false);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    21,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).not.` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a !== b).resolves.not.toBe(true);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    30,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).resolves.` + equalityMatcher + `(b);`
						}),
					},
				},
			},
			{
				Code: `expect(a !== b).resolves.not.toBe(false);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{
						MessageId: "useEqualityMatcher",
						Line:      1,
						Column:    30,
						Suggestions: expectSuggestions(func(equalityMatcher string) string {
							return `expect(a).resolves.not.` + equalityMatcher + `(b);`
						}),
					},
				},
			},
		},
	)
}
